File: dashboard/server/board/sources/ashby.js
import { COMPANIES } from "./companies.js";

/**
 * Ashby public job board API. Same bargain as Greenhouse: exact
 * `publishedAt`, a direct `applyUrl`, no auth, no ban risk.
 */

const API = "https://api.ashbyhq.com/posting-api/job-board";

function normalize(job, slug) {
  return {
    posting_id: `ashby:${slug}:${job.id}`,
    title: job.title,
    company: slug,
    location: job.location || null,
    // applyUrl lands on the form itself; jobUrl is the description page.
    job_url: job.applyUrl || job.jobUrl,
    posted_at: job.publishedAt ? job.publishedAt.slice(0, 10) : null,
    posted_at_precise: job.publishedAt
      ? new Date(job.publishedAt).toISOString()
      : null,
    applicants: null,
  };
}

export const ashbySource = {
  name: "ashby",

  async fetchRecent({ windowSeconds }) {
    const cutoff = Date.now() - windowSeconds * 1000;
    const out = [];

    for (const slug of COMPANIES.ashby) {
      try {
        const res = await fetch(`${API}/${slug}`);
        if (!res.ok) {
          console.warn(`[board] ashby/${slug} -> HTTP ${res.status}, skipped`);
          continue;
        }
        const { jobs = [] } = await res.json();

        for (const job of jobs) {
          // isListed false means pulled from the public board.
          if (job.isListed === false) continue;
          const p = normalize(job, slug);
          if (!p.posted_at_precise) continue;
          if (Date.parse(p.posted_at_precise) < cutoff) continue;
          out.push(p);
        }
      } catch (err) {
        console.warn(`[board] ashby/${slug} failed: ${err.message}`);
      }
    }

    return out;
  },

  async enrich(posting) {
    return posting;
  },

  get sliceCount() {
    return COMPANIES.ashby.length;
  },
};

File: dashboard/server/board/sources/ashby.py
"""Ashby public job board API. Same bargain as Greenhouse: exact
`publishedAt`, a direct `applyUrl`, no auth, no ban risk.
"""

import logging

from board.sources.companies import companies
from board.sources.http import client
from board.timeutil import iso_from_ms, now_ms, parse_ms

log = logging.getLogger("board")

API = "https://api.ashbyhq.com/posting-api/job-board"


def _normalize(job: dict, slug: str) -> dict:
    published = job.get("publishedAt")
    published_ms = parse_ms(published)
    return {
        "posting_id": f"ashby:{slug}:{job['id']}",
        "title": job.get("title"),
        "company": slug,
        "location": job.get("location") or None,
        # applyUrl lands on the form itself; jobUrl is the description page.
        "job_url": job.get("applyUrl") or job.get("jobUrl"),
        "posted_at": published[:10] if published else None,
        "posted_at_precise": iso_from_ms(published_ms) if published_ms is not None else None,
        "applicants": None,
    }


class AshbySource:
    name = "ashby"

    async def fetch_recent(self, window_seconds: int) -> list[dict]:
        cutoff = now_ms() - window_seconds * 1000
        out = []
        for slug in companies("ashby"):
            try:
                res = await client.get(f"{API}/{slug}")
                if res.status_code != 200:
                    log.warning("[board] ashby/%s -> HTTP %d, skipped", slug, res.status_code)
                    continue
                for job in res.json().get("jobs", []):
                    # isListed false means pulled from the public board.
                    if job.get("isListed") is False:
                        continue
                    p = _normalize(job, slug)
                    if not p["posted_at_precise"]:
                        continue
                    if parse_ms(p["posted_at_precise"]) < cutoff:
                        continue
                    out.append(p)
            except Exception as err:
                log.warning("[board] ashby/%s failed: %s", slug, err)
        return out

    async def enrich(self, posting: dict) -> dict:
        return posting

    @property
    def slice_count(self) -> int:
        return len(companies("ashby"))


ashby_source = AshbySource()
